pkg/ingester/stream_rate_calculator.go

```diff
@@ -66,31 +71,79 @@ func (c *StreamRateCalculator) updateLoop() {
 }
 
 func (c *StreamRateCalculator) updateRates() {
-	rates := make([]logproto.StreamRate, 0, c.size)
+	rates := c.rates()
+	samples := c.currentSamplesPerTenant()
+	samples = updateSamples(rates, samples)
+
+	updatedRates := make([]logproto.StreamRate, 0, c.size)
+	for _, tenantRates := range samples {
+		for _, streamRates := range tenantRates {
+			updatedRates = append(updatedRates, streamRates)
+		}
+	}
+
+	c.rateLock.Lock()
+	defer c.rateLock.Unlock()
+
+	c.allRates = updatedRates
+}
+
+func (c *StreamRateCalculator) rates() []logproto.StreamRate {
+	c.rateLock.RLock()
+	defer c.rateLock.RUnlock()
+
+	return append([]logproto.StreamRate(nil), c.allRates...)
+}
+
+func (c *StreamRateCalculator) currentSamplesPerTenant() map[string]map[uint64]logproto.StreamRate {
+	rates := make(map[string]map[uint64]logproto.StreamRate, c.size)
 
 	for i := 0; i < c.size; i++ {
 		c.locks[i].Lock()
 
-		tenantRates := c.samples[i]
-		for _, tenant := range tenantRates {
-			for _, streamRate := range tenant {
-				rates = append(rates, logproto.StreamRate{
+		for tenantID, tenant := range c.samples[i] {
+			existingRates := ratesForTenant(rates, tenantID)
```

Comment:
sorry, I found this part confusing; do you mind clarifying why do we have to check the rates variable here? I think the confusion is because:
- `rates` is a new variable populated from the `existingRates`
- the `existingRates` variable is populated here from the `rates`

which at the end looks like a cyclic dependency :thinking:

Reply:
This is because the data structure is now a `map[string]map[uint64]logproto.StreamRate` the 0 value of a map is `nil` so we have to do all this to make sure the inner map is initialized: https://github.com/grafana/loki/blob/82d838a9cc1c485084fe51e844d8bde0ed129bcd/pkg/ingester/stream_rate_calculator.go#L205